diff --git a/Makefile.release b/Makefile.release
index 28570aa2..5bb5c6d7 100644
--- a/Makefile.release
+++ b/Makefile.release
@@ -28,7 +28,7 @@ MINIFIER := $(BINDIR)minify_js_scripts \
 		--extjs-auto-deps \
 		--extjs-core-bundle $(MEDSRC)/assets/extjs/ext-dev.js \
 		$(MANIFEST)
-YUICOMPRESSOR := $(BINDIR)yuicompressor
+YUICOMPRESSOR := yuicompressor
 SOL_INIT = $(SOLSRC)/__init__.py
 
 .PHONY: bump-beta-version
