# -*- coding: utf-8 -*-
from setuptools import setup

packages = \
['gene_outlier_detection']

package_data = \
{'': ['*']}

install_requires = \
['arviz>=0.4.0,<0.5.0',
 'black>=19.10b0,<20.0',
 'click>=7.0,<8.0',
 'matplotlib>=3.0,<4.0',
 'numpy>=1.16,<2.0',
 'pandas>=0.24.1,<0.25.0',
 'pymc3>=3.8,<4.0',
 'scipy>=1.2,<2.0',
 'seaborn>=0.9.0,<0.10.0',
 'sklearn>=0.0.0,<0.0.1',
 'tables>=3.5,<4.0',
 'theano>=1.0.4,<2.0.0',
 'tqdm>=4.31,<5.0']

entry_points = \
{'console_scripts': ['outlier-detection = '
                     'gene_outlier_detection.meta_runner:cli']}

setup_kwargs = {
    'name': 'gene-outlier-detection',
    'version': '1.0',
    'description': 'A Bayesian model for identifying gene expression outliers for individual single samples (N-of-1) when compared to a cohort of background datasets.',
    'long_description': '# N-of-1 Gene Outlier Detection\n### For RNA-seq Expression Data\n\n[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/ambv/black) [![Build Status](https://travis-ci.com/jvivian/gene-outlier-detection.svg?branch=master)](https://travis-ci.com/jvivian/gene-outlier-detection) [![Coverage Status](https://coveralls.io/repos/github/jvivian/gene-outlier-detection/badge.svg)](https://coveralls.io/github/jvivian/gene-outlier-detection)\n\nThis package identifies outliers for gene expression data by building a consensus distribution from background datasets that are informed by an N-of-1 sample. \nSee [Model Explanation](#model-explanation) for more information, our paper in [JCO](https://ascopubs.org/doi/10.1200/CCI.19.00095), or our [preprint](https://www.biorxiv.org/content/early/2019/06/06/662338.full.pdf). \n\nThe model has undergone significant speed improvements and is no longer identical the model in the paper, but produces results that are essentially identical. Another benefit is the prior hyperparameters for each gene/dataset combination are now directly shared in the model whereas before they were approximated due to runtime.\n\n<p align="center"> \n<img src="/imgs/Experimental-Protocol.png" height="50%" width="50%">\n</p>\n\nThis workflow takes gene expression data as input and outputs the following:\n\n    SAMPLE_UUID\n    ├── _info\n    │\xa0\xa0 ├── _gelman-rubin.tsv\n    │\xa0\xa0 ├── _pearson_correlations.txt\n    │\xa0\xa0 ├── _pval_runs.tsv\n    │\xa0\xa0 └── _run_info.tsv\n    ├── model.pkl\n    ├── pvals.tsv\n    ├── ranks.tsv\n    ├── traceplot.png\n    ├── weights.png\n    └── weights.tsv\n- The **_info** subdirectory contains secondary information about the model run\n    - **_gelman-rubin.tsv** - TSV of [Gelman-Rubin diagnostic](https://docs.pymc.io/api/diagnostics.html#pymc3.diagnostics.gelman_rubin) for every model parameter, including a median across parameters which should be about ~1.0.\n    - **_pearson_correlations.txt** - Single column list of the Pearson correlation of gene p-values between runs (unless `-d` is provided)\n    - **_pval_runs.tsv** - Table of gene p-values as background datasets are added\n    - **_run_info.tsv** - TSV of software parameters for reproducibility and model runtime\n- **model.pkl** — A python pickle of the [PyMC3](https://docs.pymc.io) `model` and `trace`. Model must be run with `--save-model` flag and can be retrieved via\n```python\nimport pickle\nwith open(pkl_path, \'rb\') as buff:\n    data = pickle.load(buff)\nmodel, trace = data[\'model\'], data[\'trace\']\n```\n- **pvals.tsv** — P-values for all genes the model was trained on\n- **ranks.tsv** — The median rank of all groups as measured by pairwise euclidean distance\n- **traceplot.png** — Traceplot from PyMC3 linear model coefficients and model error\n- **weights.png** — Boxplot of model weights for all background datasets\n- **weights.tsv** — Mean and SD of model weights as related to background datasets\n\n# Quickstart\n1. Install\n```bash\npip install --pre gene-outlier-detection\n```\n2. Download the prerequisite [inputs](https://github.com/jvivian/gene-outlier-detection/wiki/Model-Inputs)\n3. Run the model\n```bash\noutlier-detection --sample /data/tumor.hd5 \\\n        --background /data/gtex.hd5 \\\n        --name TCGA-OR-A5KV-01 \\\n        --gene-list /data/drug-genes.txt \\\n        --col-skip 5\n```\n\n# Dependencies and Installation\n\nThis workflow has been tested on ubuntu 18.04 and Mac OSX, but should also run on other unix based systems or under an Anaconda installation.\n\n1. Python 3.6\n2. [Docker](https://docs.docker.com/install/) if using the Docker version or Toil workflow version\n3. HDF5 library (if inputs are in HDF5 format)\n    1. `conda install -c anaconda hdf5`\n4. C++ / GCC compiler for PyMC3\'s Theano\n    1. You have a couple options\n        1. `conda install theano`\n        1. `apt-get update && apt-get install -y libhdf5-serial-dev build-essential gcc`\n    \nYou _may_ need to modify your `~/.theanorc` to support larger bracket depth for this model.\n```\n[gcc]\ncxxflags = -fbracket-depth=1024\n```\n\n# Model Explanation\n\nIn the following plate notation, G stands for Gene, and D stands for Dataset (background dataset).\n\n<p align="center"> \n<img src="/imgs/Plate-Notation.png" height="50%" width="50%">\n</p>\n\nThe core of our method is a Bayesian statistical model for the N-of-1 sample’s gene expression.\nThe model implicitly assumes that the sample’s gene expression can be approximated by a\nconvex mixture of the gene expression of the background datasets. The coefficients of this\nmixture are shared across genes, much like a linear model in which each data point is the\nvector of expressions for a gene across the background datasets. In addition, we model\nexpression for each gene from each background dataset as a random variable itself. This allows\nus to incorporate statistical uncertainty from certain background sets’ small sample size directly\nin the model without drowning out any background set’s contribution through pooling \n\nThe model can be explored using Markov chain Monte Carlo (MCMC) to obtain samples\nfor _y_ (per gene) that approximate its posterior distribution. If we have an observed expression value for a\ngene of interest (from the N-of-1 cancer sample), we can compare it to the sampled values. The\nproportion of sampled values for _y_ that are greater (or lesser) than the observed value is an\nestimate of the posterior predictive p-value for this expression value. The posterior predictive\np-value can be seen as a measure of how much of an outlier the expression is given the\nexpectations of the comparison set\n\nThis model has been drastically improved for speed by leveraging PyMC3\'s vectorization \napproach and replacing the problematic student-T sampling with a Normal distribution. \nAnother benefit is there is no longer any computational "hack" of pre-fitting the \nstudent-T distributions as the expression values in the background are directly shared \nbetween two R.V.s that model expression for each gene/dataset combination. \nA graph of the new model, for 125 genes and 3 datasets with 48,500 values in the \nbackground dataset is shown below:\n\n<p align="center"> \n<img src="/imgs/new-model.png" height="50%" width="50%">\n</p>\n\n# Defining Inputs\n\nThe model requires two **Sample** by **Gene** matrices, one containing the N-of-1 sample and one containing samples to use as the background comparison set. \nThey *should* both contain the same set of genes and the background dataset must contain at least one metadata column with labels that differentiate groups (e.g. tissue, subtype, experiment, etc). Metadata columns need to be the **left-most** columns in the dataframe followed by all genes. This allows a simple heuristic (`--col-skip`) to designate where the genes start without needing the user to provide an additional input file.\n\nHere, `tissue` is one group that is used to differentiate the samples. We would run with `--col-skip=5` as there are 5 metadata columns before the start of the genes.\n\n<p align="center"> \n<img src="/imgs/example-dataframe.png" height="50%" width="50%">\n</p>\n\n# Docker Container\n\nA Docker container containing the program can be executed as follows:\n\n```bash\ndocker run --rm -v $(pwd):/data jvivian/gene-outlier-detection \\\n        outlier-model \\\n        --sample /data/inputs/tumor.hd5 \\\n        --background /data/inputs/gtex.hd5 \\\n        --name=TCGA-OR-A5KV-01 \\\n        --gene-list /data/inputs/drug-genes.txt \\\n        --out-dir /data/outputs/ \\\n        --col-skip=5\n```\n\n\n# Arguments\nExplanation of arguments used when running the program.\n\n- `--sample`\n    - A path to the matrix (.tsv / .hd5) that contains the sample.\n- `--background`\n    - A path to the matrix that contains the background datasets and at least one column at the beginning used as the label vector for different groups in the background.\n- `--name`\n    - Name of row in the sample matrix that corresponds to the desired sample to run.\n- `--out-dir`\n    - Output directory\n- `--gene-list`\n    - Single column file of genes for the model to train on and calcluate p-values for. After ~100-200 genes, it is better to split the genes into batches and run in parallel.\n- `--col-skip`\n    - Number of metadata columns to skip in background matrix. All columns after this number of columns should be genes with expression values. \n- `--group`\n    - Name of the categorical column vector in the background matrix which distinguishes the different background datasets.\n- `--num-backgrounds`  \n    - Maximum number of background categorical groups to include in the model training. Model will run, starting with one background dataset, and iteratively adding more until the p-values converge or `num-backgrounds` is met.\n- `--max-genes`\n    - Maximum number of genes to run. I.e. if a gene list is provided, how many additional genes using ANOVA. Useful for improving beta coefficients if gene list does not contain enough tissue-specific genes. It is recommended to run the model with `max-genes` set to a minimum of 10-20 more genes than exist in the `--gene-list`.\n- `--pval-convergence-cutoff`\n    - P-value Pearson correlation cutoff to stop adding additional background datasets.\n- `--num-training-genes`gi\n    - If gene-list is empty, will use ANOVA to choose gene set. Not typically useful outside of testing.\n- `--tune`\n    - Number of tuning steps to start MCMC sampling process. Default is 500, but 750 or 1,000 may be useful in circumstances where the model is having difficulty converging.\n- `--disable-iter`\n    - This flag disables iterative runs and runs one model with `--num-backgrounds`.\n- `--save-model`\n    - This flag will save a serialized version of the model/trace. Useful for debugging or inspection of all model parameters.\n    \n# Toil-version of Workflow\n\nA [Toil](https://toil.readthedocs.io/) version of the workflow is available [here](https://github.com/jvivian/gene-outlier-detection/blob/master/toil/toil-outlier-detection.py). This allows the model to be run on multiple samples at scale on a cluster or cloud computing cluster, but requires Python 2.7 and `pip install pandas toil==3.19.0`\n\nToil is now Python 3 compliant, so this method of running the program isn\'t recommended until the Toil code has been updated.\n    \n# Citation\nIf you leverage this method in your research, please cite our [JCO](https://ascopubs.org/doi/10.1200/CCI.19.00095) paper.\n```\nVivian, John, et al. "Bayesian Framework for Detecting \nGene Expression Outliers in Individual Samples." \nJCO Clinical Cancer Informatics 4 (2020): 160-170.\n```\n',
    'author': 'John Vivian',
    'author_email': 'jtvivian@gmail.com',
    'maintainer': None,
    'maintainer_email': None,
    'url': None,
    'packages': packages,
    'package_data': package_data,
    'install_requires': install_requires,
    'entry_points': entry_points,
    'python_requires': '>=3.6,<4.0',
}


setup(**setup_kwargs)
